yamux/src/stream.rs

```diff
@@ -396,21 +432,49 @@ impl AsyncRead for StreamHandle {
             return Poll::Ready(Ok(()));
         }
 
-        let n = ::std::cmp::min(buf.remaining(), self.read_buf.len());
+        if self.read_buf.is_empty() {
+            self.readable_wake = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        let mut offset = 0;
+        let mut total_read = 0;
+        let mut need_drain = false;
+        for (index, read_buf) in self.read_buf.iter_mut().enumerate() {
+            let n = ::std::cmp::min(buf.remaining(), read_buf.len());
+            if n == 0 {
+                break;
+            }
+            total_read += n;
+            if n == read_buf.len() {
+                let b = &read_buf[..n];
+                buf.put_slice(b);
+                offset = index;
+                need_drain = true;
+            } else {
+                let b = read_buf.split_to(n);
+                buf.put_slice(&b);
+            }
+        }
+
+        if need_drain {
```

Comment:
offset and need_drain can be combined to one Option var, and the loop can be simplified:

```suggestion
        let mut offset = None;
        let mut total_read = 0;
        for (index, read_buf) in self.read_buf.iter_mut().enumerate() {
            let n = buf.remaining().min(read_buf.len());
            if n == 0 {
                break;
            }
            buf.put_slice(&read_buf.split_to(n));
            if read_buf.is_empty() {
                offset = Some(index);
            }
            total_read += n;
        }

        if let Some(offset) = offset {
```

this suggestion is also applied to other